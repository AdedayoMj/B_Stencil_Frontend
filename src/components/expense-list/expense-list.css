:host {
  display: block;
}
.table-container {
  margin-top: 5rem;
  display: flex;
  justify-content: center;
  margin-bottom: 5rem;
}
table {
  width: 100%;
  border-collapse: collapse;
}

th,
td {
  padding: 8px;
  text-align: left;
  border-bottom: 1px solid #ddd;
}
th {
  color: white;
}

th {
  background-color: var(--app-button-color2);
}

.pagination {
  display: flex;
  justify-content: center;
  margin-top: 1rem;
}

.pagination button {
  margin: 0 1rem;
  padding: 0.5rem;
  border: none;
  background-color: var(--app-button-color2);
  color: white;
  cursor: pointer;
}

.pagination button:hover {
  background-color: var(--app-button-color);
}

.pagination .active {
  background-color: var(--app-button-color);
}

.actions button {
  margin-right: 0.5rem;
  padding: 0.5rem;
  border: none;
  border-radius: 4px;
  width: 4rem;
  background-color: var(--app-button-color2);
  color: white;
  cursor: pointer;
}

.actions button:hover {
  background-color: var(--app-button-color);
}
.actions .delete {
  background-color: #e41919;
}
.actions .delete:hover {
  background-color: #cc3333;
}
