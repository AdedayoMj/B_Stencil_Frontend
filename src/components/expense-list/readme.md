# expense-list



<!-- Auto Generated Below -->


## Properties

| Property   | Attribute | Description | Type            | Default     |
| ---------- | --------- | ----------- | --------------- | ----------- |
| `expenses` | --        |             | `ExpenseData[]` | `undefined` |


## Dependencies

### Used by

 - [app-root](../app-root)

### Graph
```mermaid
graph TD;
  app-root --> expense-list
  style expense-list fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
