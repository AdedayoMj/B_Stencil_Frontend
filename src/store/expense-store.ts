import { createStore } from '@stencil/store';
import { CreateExpenseData, ExpenseData } from '../types';

interface AppState {
  expenses: ExpenseData[];
  theme: string;
  showForm: boolean;
  currentId: string;
  action: string;
}

const intialState: AppState = {
  expenses: [],
  theme: 'light',
  showForm: false,
  currentId: '',
  action: 'create',
};

const apiUrl = 'https://betsson-backend-ed81ddb5eddb.herokuapp.com/expense';
// const apiUrl = process.env.API_URL;
const { state, onChange } = createStore(intialState);

const CAHCE_EXPIRATION = 60 * 5;

const setCache = (key: string, value: ExpenseData[]) => {
  const expirationTime = Date.now() + CAHCE_EXPIRATION * 1000;
  const cacheObject = { value, expirationTime };
  localStorage.setItem(key, JSON.stringify(cacheObject));
};

const getCache = (key: string) => {
  const cacheData = localStorage.getItem(key);
  if (cacheData) {
    const { value, expirationTime } = JSON.parse(cacheData);
    if (expirationTime > Date.now()) {
      return value;
    } else {
      localStorage.removeItem(key);
    }
  }
  return null;
};

export const expenseStore = {
  getState: () => state,
  subscribe: onChange,

  setExpense: (expenses: ExpenseData[]) => {
    state.expenses = expenses;
    setCache('expenses', expenses);
  },

  addExpense: (expense: ExpenseData) => {
    state.expenses = [...state.expenses, expense];
    setCache('expenses', state.expenses);
  },

  deleteExpense: (expenseId: string) => {
    state.expenses = state.expenses.filter(expense => expense._id !== expenseId);
    setCache('expenses', state.expenses);
  },
  
  updateExpense: (expenseId: string, updatedExpense: ExpenseData) => {
    state.expenses = state.expenses.map(expense => (expense._id === expenseId ? { ...expense, ...updatedExpense } : expense));
    setCache('expenses', state.expenses);
  },

  setTheme: (theme: string) => {
    state.theme = theme;
    localStorage.setItem('theme', theme);
  },
  setShowForm: (showForm: boolean) => {
    state.showForm = showForm;
  },
  setAction: (action: string) => {
    state.action = action;
  },
  setCurrentId: (id: string) => {
    state.currentId = id;
  },

  fetchExpenses: async () => {
    const cachedExpenses = getCache('expenses');
    const cachedTheme = localStorage.getItem('theme');
    
    if (cachedTheme) {
      expenseStore.setTheme(cachedTheme);
    } else {
      expenseStore.setTheme('light'); 
    }
  
    if (cachedExpenses) {
      expenseStore.setExpense(cachedExpenses);
      return;
    }

    try {
      const response = await fetch(`${apiUrl}`);
      if (!response.ok) {
        throw new Error('Failed to fetch expenses');
      }
      const data = await response.json();
      expenseStore.setExpense(data.expenses);

      setCache('expenses', data.expenses);
    } catch (error) {
      console.error('Error fetching expenses:', error);
      throw error;
    }
  },

  createExpense: async (requestData: CreateExpenseData) => {
    try {
      const response = await fetch(`${apiUrl}/createExpense`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(requestData),
      });
      if (!response.ok) {
        console.error('Error creating expense:', response.statusText);
        throw new Error('Error creating expense');
      }
      const data = await response.json();
      expenseStore.addExpense(data.expense);
    } catch (error) {
      console.error('Error creating expense:', error);
      throw error;
    }
  },

  deleteExpenseApi: async (expenseId: string) => {
    try {
      const response = await fetch(`${apiUrl}/${expenseId}`, {
        method: 'DELETE',
      });

      if (!response.ok) {
        console.error('Error deleting expense:', response.statusText);
        throw new Error('Error deleting expense');
      }
      expenseStore.deleteExpense(expenseId);
    } catch (error) {
      console.error('Error deleting expense:', error);
      throw error;
    }
  },
  updateExpenseApi: async (expenseId: string, updatedExpense: CreateExpenseData) => {
    try {
      const response = await fetch(`${apiUrl}/update/${expenseId}`, {
        method: 'PATCH',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(updatedExpense),
      });
      if (!response.ok) {
        console.error('Error updating expense:', response.statusText);
        throw new Error('Error updating expense');
      }

      expenseStore.updateExpense(expenseId, updatedExpense);
    } catch (error) {
      console.error('Error updating expense:', error);
      throw error;
    }
  },

  getExpenseById: (expenseId: string): ExpenseData | undefined => {
    return state.expenses.find(expense => expense._id === expenseId);
  },
};
