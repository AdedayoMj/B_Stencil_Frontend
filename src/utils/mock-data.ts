import { ExpenseData } from '../types';

export const mockExpenses: ExpenseData[] = [
  { date: '2023-07-01', description: 'Expense 1', amount: 10 },
  { date: '2023-07-02', description: 'Expense 2', amount: 20 },
  { date: '2023-07-03', description: 'Expense 3', amount: 30 },
  { date: '2023-07-04', description: 'Expense 4', amount: 40 },
  { date: '2023-07-05', description: 'Expense 5', amount: 50 },
  { date: '2023-07-06', description: 'Expense 6', amount: 60 },
  { date: '2023-07-07', description: 'Expense 7', amount: 70 },
];
