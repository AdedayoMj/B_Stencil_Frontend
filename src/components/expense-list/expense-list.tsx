import { Component, Host, Prop, State, h } from '@stencil/core';
import { ExpenseData } from '../../types';
import { expenseStore } from '../../store/expense-store';

@Component({
  tag: 'expense-list',
  styleUrl: 'expense-list.css',
  shadow: true,
})
export class ExpenseList {
  @Prop() expenses: ExpenseData[];

  @State() currentPage = 1;
  @State() itemsPerPage = 5;

  handlePageChange(pageNumber: number) {
    this.currentPage = pageNumber;
  }

  selectExpense(expense: ExpenseData) {
    expenseStore.setAction('update');
    expenseStore.setCurrentId(expense._id);
    expenseStore.setShowForm(true);
  }

  async deleteExpense(expenseId: string) {
    try {
      await expenseStore.deleteExpenseApi(expenseId);
      await expenseStore.fetchExpenses();
    } catch (error) {
      console.error('Error deleting expense:', error);
    }
  }

  render() {
    const { expenses, currentPage, itemsPerPage } = this;

    if (!expenses || expenses.length === 0) {
      return <h1>No expenses</h1>;
    }

    const sortedExpenses = [...expenses].sort((a, b) => {
      const dateA = new Date(a.date);
      const dateB = new Date(b.date);
      return dateB.getTime() - dateA.getTime();
    });

    const startIndex = (currentPage - 1) * itemsPerPage;
    const endIndex = startIndex + itemsPerPage;
    const currentExpenses = sortedExpenses.slice(startIndex, endIndex);

    const totalPages = Math.ceil(sortedExpenses.length / itemsPerPage);
    return (
      <Host>
        <div class="table-container">
          <div>
            <table>
              <thead>
                <tr>
                  <th>Date</th>
                  <th>Expense Details</th>
                  <th>Amount</th>
                  <th>Actions</th>
                </tr>
              </thead>
              <tbody>
                {currentExpenses.map(expense => (
                  <tr key={expense._id}>
                    <td>{expense.date}</td>
                    <td>{expense.description}</td>
                    <td>{expense.amount}</td>
                    <td class="actions">
                      {' '}
                      <button onClick={() => this.selectExpense(expense)}>Edit</button>
                      <button class="delete" onClick={() => this.deleteExpense(expense._id)}>
                        Delete
                      </button>
                    </td>
                  </tr>
                ))}
              </tbody>
            </table>
            <div class="pagination">
              <button disabled={currentPage === 1} onClick={() => this.handlePageChange(currentPage - 1)}>
                Prev
              </button>
              <span>{`${currentPage}/${totalPages}`}</span>
              <button disabled={currentPage === totalPages} onClick={() => this.handlePageChange(currentPage + 1)}>
                Next
              </button>
            </div>
          </div>
        </div>
      </Host>
    );
  }
}
