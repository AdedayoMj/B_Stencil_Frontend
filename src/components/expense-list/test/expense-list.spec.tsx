import { newSpecPage } from '@stencil/core/testing';
import { ExpenseList } from '../expense-list';
import { expenseStore } from '../../../store/expense-store';
import { mockExpenses } from '../../../utils/mock-data';

describe('ExpenseList', () => {
  it('should render expense list correctly', async () => {
    jest.spyOn(expenseStore, 'fetchExpenses').mockImplementation(async () => {
      expenseStore.setExpense(mockExpenses);
    });
    const page = await newSpecPage({
      components: [ExpenseList],
      html: `<expense-list></expense-list>`,
      supportsShadowDom: true,
    });

    expect(page.root).toBeTruthy();
    expect(page.root).toMatchSnapshot();
  });

  it('should display "No expenses" message when no expenses are provided', async () => {
    const expenses = [];
    jest.spyOn(expenseStore, 'fetchExpenses').mockImplementation(async () => {
      expenseStore.setExpense(expenses);
    });

    const page = await newSpecPage({
      components: [ExpenseList],
      html: `<expense-list></expense-list>`,
      supportsShadowDom: true,
    });

    const noExpensesMessage = page.root.shadowRoot.querySelector('h1');
    expect(noExpensesMessage).toBeTruthy();
    expect(noExpensesMessage.textContent).toBe('No expenses');
  });

  it('should handle pagination correctly', async () => {
    jest.spyOn(expenseStore, 'fetchExpenses').mockImplementation(async () => {
      expenseStore.setExpense(mockExpenses);
    });
    const page = await newSpecPage({
      components: [ExpenseList],
      html: `<expense-list></expense-list>`,
    });

    await page.waitForChanges();

    const expenseList = page.rootInstance as ExpenseList;

    expect(page.root).toMatchSnapshot();
    expect(expenseList.currentPage).toBe(1);

    expenseList.handlePageChange(2);

    await page.waitForChanges();

    expect(page.root).toMatchSnapshot();
    expect(expenseList.currentPage).toBe(2);
  });
});
