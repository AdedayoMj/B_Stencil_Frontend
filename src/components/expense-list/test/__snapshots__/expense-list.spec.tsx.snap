// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpenseList should handle pagination correctly 1`] = `
<expense-list>
  <mock:shadow-root>
    <h1>
      No expenses
    </h1>
  </mock:shadow-root>
</expense-list>
`;

exports[`ExpenseList should handle pagination correctly 2`] = `
<expense-list>
  <mock:shadow-root>
    <h1>
      No expenses
    </h1>
  </mock:shadow-root>
</expense-list>
`;

exports[`ExpenseList should render expense list correctly 1`] = `
<expense-list>
  <mock:shadow-root>
    <h1>
      No expenses
    </h1>
  </mock:shadow-root>
</expense-list>
`;
